### PR TITLE
Improve performance of JSDoc tag utilities

### PR DESCRIPTION
**Pull request by andy-ms**
We currently have no good performance regression tests of large `--checkJs` projects, since it's a new feature. Previously the process for something like `getJSDocReturnTag` looked like this:

* Look at the cache. If it is `undefined`, either it's not in the cache yet, or we already tried getting jsdoc tags and there were none.
* Iterate over all JSDocs and collect an array of comments / tags.
* Then flatMap over it:
	- At each doc comment, create a new array that filters its tags of the correct kind.
* Get `firstOrUndefined` of that.

This changes it to:

* Look at the cache. If it is `null`, then we've already determined that this node has no jsdoc tags.
* If it is `undefined`, iterate and get tags as before.
* Iterate over the (cached) array of tags looking for one of kind `JSDocReturnTag`.

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -1437,39 +1437,42 @@ namespace ts {
...
-        return flatMap(getJSDocs(node), doc =>
-            doc.kind === SyntaxKind.JSDocComment
-                ? filter((doc as JSDoc).tags, tag => tag.kind === kind)
-                : doc.kind === kind && doc);
+        return !!getFirstJSDocTag(node, SyntaxKind.JSDocParameterTag);
     }
 
-    function getFirstJSDocTag(node: Node, kind: SyntaxKind): JSDocTag {
-        return node && firstOrUndefined(getJSDocTags(node, kind));
+    function getFirstJSDocTag(node: Node, kind: SyntaxKind) {
```
Comment (yuit) [inline src/compiler/utilities.ts]: nit: return type

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -1437,39 +1437,42 @@ namespace ts {
             (<JSDocFunctionType>node).parameters[0].type.kind === SyntaxKind.JSDocConstructorType;
     }
 
-    export function getCommentsFromJSDoc(node: Node): string[] {
-        return map(getJSDocs(node), doc => doc.comment);
-    }
-
     export function hasJSDocParameterTags(node: FunctionLikeDeclaration | SignatureDeclaration) {
```
Comment (yuit) [inline src/compiler/utilities.ts]: nit: return type

File: src/compiler/utilities.ts [hunk 3]
```diff
@@ -1437,39 +1437,42 @@ namespace ts {
...
 
-    function getFirstJSDocTag(node: Node, kind: SyntaxKind): JSDocTag {
-        return node && firstOrUndefined(getJSDocTags(node, kind));
+    function getFirstJSDocTag(node: Node, kind: SyntaxKind) {
+        const tags = getJSDocTags(node);
+        return tags && find(tags, doc => doc.kind === kind);
     }
 
-    export function getJSDocs(node: Node): (JSDoc | JSDocTag)[] {
+    export function getAllJSDocs(node: Node): Array<JSDoc | JSDocTag> | undefined {
```
Comment (yuit) [inline src/compiler/utilities.ts]: Q: why did we use `Array<JSDoc | JSDocTag>` instead ` (JSDoc | JSDocTag)[]` ?

File: src/compiler/utilities.ts [hunk 4]
```diff
@@ -1437,39 +1437,42 @@ namespace ts {
...
             return [node.parent];
         }
+        return getJSDocCommentsAndTags(node);
+    }
 
-        let cache: (JSDoc | JSDocTag)[] = node.jsDocCache;
-        if (!cache) {
-            getJSDocsWorker(node);
-            node.jsDocCache = cache;
+    export function getJSDocTags(node: Node): JSDocTag[] | undefined {
```
Comment (yuit) [inline src/compiler/utilities.ts]: Q: why this function get export?

Comment (andy-ms) [inline src/compiler/utilities.ts]: It's needed in `getJsDocTagsFromDeclarations`.

**Pull request merged**